Fixed ambiguity when using Should on a JsonNode derived class

JSkimming opened the pull request:
The first commit demonstrates the issue; the second commit proposes a solution.

The third commit makes the `JsonNodeAssertions` generic, which propagates the type safety of the `Subject`.

Fixes #3103

## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [X] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [X] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/main/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/main/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/main/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/main/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/main/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

## CONTRIBUTOR LICENSE GRANT

By submitting this contribution, the contributor hereby irrevocably grants to the project owners and maintainers a perpetual, worldwide, royalty-free, irrevocable license to use, reproduce, modify, distribute, sublicense, and create derivative works of the contribution for any purpose and under any terms, including proprietary licensing.

The contributor waives any moral rights in the contribution to the extent permitted by law and agrees not to assert any claim of authorship or control over the contribution. The contributor represents that they are the sole author of the contribution and that it is provided free of any third-party claims.

The contributor understands and agrees that the maintainers may, at their sole discretion, use, license, or redistribute the contribution as part of any work and under any terms they choose, without further permission or attribution.

* [X] I have read the Contributor License Grant and accept the conditions
* [x] I'm interested in a free license for Fluent Assertions and will share my email address through sales@xceed.com

dennisdoomen commented on Src/FluentAssertions/Specialized/JsonNodeAssertions.cs:
```diff
@@ -13,10 +13,11 @@
 namespace FluentAssertions.Specialized;
 
-public class JsonNodeAssertions : ReferenceTypeAssertions<JsonNode, JsonNodeAssertions>
+public class JsonNodeAssertions<T> : ReferenceTypeAssertions<T, JsonNodeAssertions<T>>
```
Since you've changed the public API, you also need to run `AcceptApiChanges.ps1` to update the snapshot tests.

JSkimming replied:
Done.

jnyrup commented on Src/FluentAssertions/AssertionExtensions.cs:
```diff
@@ -913,9 +913,10 @@ public static TaskCompletionSourceAssertions Should(this TaskCompletionSource tc
     [Pure]
-    public static JsonNodeAssertions Should([NotNull] this JsonNode jsonNode)
+    public static JsonNodeAssertions<T> Should<T>([NotNull] this T jsonNode)
+        where T : JsonNode
     {
-        return new JsonNodeAssertions(jsonNode, AssertionChain.GetOrCreate());
+        return new JsonNodeAssertions<T>(jsonNode, AssertionChain.GetOrCreate());
     }
```
We have historically avoided generic methods in this class, as we can only have a single `Should<T>(T value)` no matter what generic constraints one have.

But we can extract them into other classes, which we did for these types:

[EnumAssertionsExtensions](https://github.com/fluentassertions/fluentassertions/blob/main/Src/FluentAssertions/EnumAssertionsExtensions.cs)
[DataSetAssertionExtensions](https://github.com/fluentassertions/fluentassertions.datasets/blob/main/Src/FluentAssertions.DataSets/DataSetAssertionExtensions.cs)
[DataTableAssertionExtensions](https://github.com/fluentassertions/fluentassertions.datasets/blob/main/Src/FluentAssertions.DataSets/DataTableAssertionExtensions.cs)
[DataRowAssertionExtensions](https://github.com/fluentassertions/fluentassertions.datasets/blob/main/Src/FluentAssertions.DataSets/DataRowAssertionExtensions.cs)

So I propose we extract it into a new class `JsonAssertionExtensions`.

```cs
#if NET6_0_OR_GREATER

using System.Linq;
using System.Text.Json.Nodes;
using FluentAssertions.Collections;
using FluentAssertions.Execution;
using FluentAssertions.Specialized;
using JetBrains.Annotations;

namespace FluentAssertions;

/// <summary>
/// Contains extension methods for custom assertions in unit tests.
/// </summary>
public static class JsonAssertionExtensions
{
    /// <summary>
    /// Returns an object that provides various assertion APIs that act on a <see cref="JsonNode"/>.
    /// </summary>
    [Pure]
    public static JsonNodeAssertions<T> Should<T>([NotNull] this T jsonNode)
        where T : JsonNode
    {
        return new JsonNodeAssertions<T>(jsonNode, AssertionChain.GetOrCreate());
    }

    /// <summary>
    /// Return an object that provides various assertion APIs that treat a <see cref="JsonArray"/>
    /// as a collection of <see cref="JsonNode"/>s.
    /// </summary>
    [Pure]
    public static GenericCollectionAssertions<JsonNode> Should([NotNull] this JsonArray jsonArray)
    {
        return new GenericCollectionAssertions<JsonNode>(jsonArray?.ToArray(), AssertionChain.GetOrCreate());
    }
}

#endif
```

dennisdoomen replied:
> We have historically avoided generic methods in this class, as it can only have a single Should<T>(T value) no matter what generic constraints or return type it has.

I was also surprised that this PR worked in the first place, but somehow it did.

JSkimming commented on Tests/FluentAssertions.Equivalency.Specs/JsonNodeSpecs.cs:
```diff
@@ -546,6 +546,16 @@ public void Can_treat_a_null_json_array_as_a_collection()
         array.Should().BeNull();
     }
+
+    [Fact]
+    public void Can_assert_json_object_instances()
+    {
+        // Arrange
+        JsonObject jsonObject = null;
+
+        // Act & Assert
+        jsonObject.Should().BeNull();
+    }
```
I've just added a `JsonValue` equivalent test, and a `JsonNode` one for completeness.

The `JsonNode` test is superfluous, given that other tests in the class are based on it. I'm happy to remove it if that's your preference.

JSkimming commented:
Hi All.

I've incorporated your feedback. Please let me know what you think.

I've also updated the documentation and the release notes and checked the spelling. I hope that all looks OK.

Finally, I've changed the name of the PR, as the compile error wasn't limited to the `BeNull()` method, but any use of `Should()` on a `JsonNode` derived class.

I'm now off to watch a movie. Perhaps some classic horror 🤔, maybe something from John Carpenter.

jnyrup commented:
Thanks for the contribution and enjoy the movie. 
I've been rewatching Twin Peaks tonight.

dennisdoomen commented:
Thanks for this fix @JSkimming 

I've been watching Tron Legacy with my son last night. It's from 2010 and he's from 2012 ;-)

JSkimming commented:
Thanks for responding so quickly.

I ended up watching From Dusk Till Dawn. I hope your son enjoyed Tron Legacy @dennisdoomen. I've taken my kids on a classic movie journey. Enjoy it.
